OrbitCore/FunctionFramepointerValidator.cpp

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "FunctionFramepointerValidator.h"
+
+#include <capstone/capstone.h>
+#include <iostream>
+
+FunctionFramepointerValidator::FunctionFramepointerValidator(
+    csh handle, const uint8_t* code, size_t code_size) {
+  handle_ = handle;
+  instructions_ = nullptr;
+  instructions_count_ =
+      cs_disasm(handle, code, code_size, 0, 0, &instructions_);
+}
+
+FunctionFramepointerValidator::~FunctionFramepointerValidator() {
+  cs_free(instructions_, instructions_count_);
+}
+
+bool FunctionFramepointerValidator::isCallInstruction(
+    const cs_insn& instruction) {
+  for (uint8_t i = 0; i < instruction.detail->groups_count; i++) {
+    if (instruction.detail->groups[i] == X86_GRP_CALL) return true;
+  }
+  return false;
+}
+
+bool FunctionFramepointerValidator::isMovInstruction(
+    const cs_insn& instruction) {
+  return instruction.id == X86_INS_MOV || instruction.id == X86_INS_MOVQ;
+}
+
+bool FunctionFramepointerValidator::isBasePointer(uint16_t reg) {
+  return reg == X86_REG_BP || reg == X86_REG_EBP || reg == X86_REG_RBP;
+}
+
+bool FunctionFramepointerValidator::isStackPointer(uint16_t reg) {
+  return reg == X86_REG_SP || reg == X86_REG_ESP || reg == X86_REG_RSP;
+}
+
+// we will actually only check, whether there is a correct epilogue
+bool FunctionFramepointerValidator::validateEpilogue() {
+  bool found_epilogue = false;
+  for (size_t i = 0; i < instructions_count_ - 1; i++) {
+    if (isMovInstruction(instructions_[i]) &&
```

Comment:
do you check if there is ret after mov and pop?

Reply:
In capstone there is: 
```// all return instructions
X86_GRP_RET```

So it should be safe to use this group, as done for calls.

But I don't get you comment with pop rip. Could you elaborate?

Reply:
if you have return address on the top of the stack pop rip is effectively a ret. The question is if amd64 lets regular instructions modify rip (or directly assign something to rip), for example arm32 lets you do that.

From what I understand amd64 does not allow that so we should be fine with checking RET

Reply:
Ah thanks for clarifying. Well lets implement the check for RET, we would only produce false positive warnings if pop rip would be possible though.